src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -5388,6 +5351,63 @@ static TypeWithState calculateResultType(TypeWithAnnotations targetTypeWithNulla
                 var resultType = TypeWithState.Create(targetType, resultState);
                 return resultType;
             }
+
+            static NullableFlowState getReferenceConversionResultState(TypeWithAnnotations targetType, TypeWithState operandType)
+            {
+                var state = operandType.State;
+                switch (state)
+                {
+                    case NullableFlowState.MaybeNull:
+                        if (operandType.Type?.IsTypeParameterDisallowingAnnotation() != true &&
```

Comment:
>IsTypeParameterDisallowingAnnotation [](start = 46, length = 36)

This is not quite the right test, as we want to permit type parameters constrained to `int?`, which disallow annotations, to be treated as known able to hold null.  See `RequiresSafetyWarningWhenNullIntroduced` for what I believe the correct test is.

Reply:
It looks like the check for `!type.IsNullableTypeOrTypeParameter` in the `RequiresSafetyWarningWhenNullIntroduced` methods do not affect any tests in `NullableReferenceTypesTests`.

And this particular local function is for reference conversions only.

---
In reply to: [347688703](https://github.com/dotnet/roslyn/pull/39778#discussion_r347688703) [](ancestors = 347688703)